Fix a handling of dlsym(RTLD_NEXT) without loading libGL. Closes #1463

asimiklit opened the pull request:
* The original fix leads to infinite recursion in GLX, EGL hooks.
  Due to `dlopen` function override in `librenderdoc.so` library the
  `Process::LoadModule` function always returns a handle of
  `librenderdoc.so` instead of requested library. So instead of
  `real function` we receive an address of the `hook` that leads
  to the infinite recursion in this `hook`.

Fixes: 66869e ("Handle dlsym(RTLD_NEXT) without loading libGL")
Signed-off-by: Andrii Simiklit <andrii.simiklit@globallogic.com>

asimiklit commented:
Hmm, both errors on `The Travis CI build failed` are related to `Could NOT find PythonInterp: Found unsuitable version "2.7.17", but required is at least "3" (found /usr/local/bin/python)`. So it isn't related to my changes)

baldurk commented:
Yeh the travis failures seem to be related to the environment updating on OSX. I've fixed them locally and will push soon, ignore that for now.

I don't think this fix is correct though. Your analysis is partly accurate but the original commit 66869e8 wasn't broken - this is clear since otherwise it wouldn't have fixed the original bug but you can try compiling it and it will work on your reproducer from the other issue. Rather the bug was introduced in 6d40bbb when wayland support was added. That's when the `eglhook.handle` was assigned since it was then needed on replay, and on replay there's no interception of dlopen so it works. The problem as you found was that it introduced the infinite loop.

I think your fix has the risk of breaking, since it could lead to registering hooks while replaying (the passthrough functions can call `EnsureRealLibraryLoaded()`), and even during capture we must not register hooks after the registration phase has ended. A better fix would be this:

```diff
diff --git a/renderdoc/driver/gl/egl_hooks.cpp b/renderdoc/driver/gl/egl_hooks.cpp
index e98fb6380..aa97237dc 100644
--- a/renderdoc/driver/gl/egl_hooks.cpp
+++ b/renderdoc/driver/gl/egl_hooks.cpp
@@ -143,7 +143,8 @@ static void EnsureRealLibraryLoaded()
     if(!handle)
       handle = Process::LoadModule("libEGL.so.1");

-    eglhook.handle = handle;
+    if(RenderDoc::Inst().IsReplayApp())
+      eglhook.handle = handle;
   }
 #endif
 }
diff --git a/renderdoc/driver/gl/glx_hooks.cpp b/renderdoc/driver/gl/glx_hooks.cpp
index 983917f88..3c2fcd758 100644
--- a/renderdoc/driver/gl/glx_hooks.cpp
+++ b/renderdoc/driver/gl/glx_hooks.cpp
@@ -76,7 +76,8 @@ static void EnsureRealLibraryLoaded()
     if(!handle)
       handle = Process::LoadModule("libGLX.so.0");

-    glxhook.handle = handle;
+    if(RenderDoc::Inst().IsReplayApp())
+      glxhook.handle = handle;
   }
 }
```

Which works for me on apitrace and your test. I haven't tested if that works/breaks for wayland but it's not currently supported so it may have broken some other time, if someone wants to make it work they can check this.

If you'd like to refactor this PR then please do, otherwise I'll commit that fix directly in a couple of days.

baldurk reviewed (changes requested):
As above, hooks cannot be registered outside of `RegisterHooks()` functions and I think the fix can be simpler.

asimiklit committed:
Fix a handling of dlsym(RTLD_NEXT) without loading libGL. Closes #1463

* The original fix leads to infinite recursion in GLX, EGL hooks.
  Due to `dlopen` function override in `librenderdoc.so` library the
  `Process::LoadModule` function always returns a handle of
  `librenderdoc.so` instead of requested library. So instead of
  `real function` we receive an address of the `hook` that leads
  to the infinite recursion in this `hook`.

Fixes: 6d40bbb ('Add experimental wayland support. Refs #853')
Tested-by: Andrii Simiklit <andrii.simiklit@globallogic.com>

asimiklit commented:
I refactored it and re-tested, works good. I just noticed that `glxhook.handle = handle;` was added by another commit, so I fixed `Fixes` tag. Thanks)
